oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizedClientRefreshedEventListener.java

```diff
@@ -0,0 +1,219 @@
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtDecoderFactory;
+import org.springframework.security.oauth2.jwt.JwtException;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * An {@link ApplicationListener} that listens for events of type
+ * {@link OAuth2AuthorizedClientRefreshedEvent} and publishes an event of type
+ * {@link OidcUserRefreshedEvent} in order to refresh an {@link OidcUser}.
+ *
+ * @author Steve Riesenberg
+ * @since 6.5
+ * @see org.springframework.security.oauth2.client.RefreshTokenOAuth2AuthorizedClientProvider
+ * @see OAuth2AuthorizedClientRefreshedEvent
+ * @see OidcUserRefreshedEvent
+ */
+public final class OidcAuthorizedClientRefreshedEventListener
+		implements ApplicationEventPublisherAware, ApplicationListener<OAuth2AuthorizedClientRefreshedEvent> {
+
+	private static final String INVALID_ID_TOKEN_ERROR_CODE = "invalid_id_token";
+
+	private static final String INVALID_NONCE_ERROR_CODE = "invalid_nonce";
+
+	private OAuth2UserService<OidcUserRequest, OidcUser> userService = new OidcUserService();
+
+	private JwtDecoderFactory<ClientRegistration> jwtDecoderFactory = new OidcIdTokenDecoderFactory();
+
+	private GrantedAuthoritiesMapper authoritiesMapper = (authorities) -> authorities;
+
+	private SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder
+		.getContextHolderStrategy();
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	@Override
+	public void onApplicationEvent(OAuth2AuthorizedClientRefreshedEvent event) {
+		if (this.applicationEventPublisher == null) {
+			return;
+		}
+
+		// The response must contain the openid scope
+		OAuth2AccessTokenResponse accessTokenResponse = event.getAccessTokenResponse();
+		if (!accessTokenResponse.getAccessToken().getScopes().contains(OidcScopes.OPENID)) {
+			return;
+		}
+
+		// The response must contain an id_token
+		Map<String, Object> additionalParameters = accessTokenResponse.getAdditionalParameters();
+		if (!StringUtils.hasText((String) additionalParameters.get(OidcParameterNames.ID_TOKEN))) {
+			return;
+		}
+
+		// The current authentication must be an OAuth2AuthenticationToken
+		Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
+		if (!(authentication instanceof OAuth2AuthenticationToken authenticationToken)
+				|| authenticationToken.getClass() != OAuth2AuthenticationToken.class) {
+			// This event listener only handles the default authentication result. If the
+			// application customizes the authentication result, then a custom event
+			// handler should be provided.
+			return;
+		}
+
+		// The current principal must be an OidcUser
+		if (!(authenticationToken.getPrincipal() instanceof OidcUser existingOidcUser)) {
```

Comment:
Testing for `OAuth2User` would allow to process OAuth2 refresh token flows without OIDC